rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+
+For the below data, Encodes and Decodes are ops/sec so higher is better while size before and after compression lower is better. This list is far from exhaustive and specifcially lacks a serde_json comparison which is unfortunate but other benches out there do include them (though obviously a different benchmark makes cross comparison difficult). 
+
+| Library            | Encode | Decode All | Decode 1 | Update 1 | Size (bytes) | Size (Zlib) |
+|--------------------|--------|------------|----------|----------|--------------|-------------|
+| **Runtime Libs**   |        |            |          |          |              |             |
+| *NoProto*          |   1057 |       1437 |    47619 |    12195 |          208 |         166 |
+| Apache Avro        |    138 |         51 |       52 |       37 |          702 |         336 |
+| FlexBuffers        |    401 |        855 |    23256 |      264 |          490 |         309 |
+| JSON               |    550 |        438 |      544 |      396 |          439 |         184 |
+| BSON               |    115 |        103 |      109 |       80 |          414 |         216 |
+| MessagePack        |    135 |        222 |      237 |      119 |          296 |         187 |
+| **Compiled Libs**  |        |            |          |          |              |             |
+| Flatbuffers        |   1046 |      14706 |   250000 |     1065 |          264 |         181 |
+| Bincode            |   5882 |       8772 |     9524 |     4016 |          163 |         129 |
+| Protobuf           |    859 |       1140 |     1163 |      480 |          154 |         141 |
+| Prost              |   1225 |       1866 |     1984 |      962 |          154 |         142 |
+
+Performance is a single aspect of a nuanced problem for us; for more perf benchmarks I'd suggest checking out [rust-serialization-benchmarks](https://github.com/erickt/rust-serialization-benchmarks).  Some of these formats (JSON for example) still leaves us with a marginal concern around both dropped metadata as well as incidental exposure of a private data format to our users. The goal here is to hopefully solve for all of these things with whatever path is chosen.
+
+While it is tempting to look at perf benchmarks and to just pick the best performing example, the practical requirements around the problem for us might give us some limitations. As was [expressed by @lukesteensen](https://discord.com/channels/742820443487993987/746070604192415834/796425792875397120) in the #development channel in discord we have three specific concerns around the choice:
+
+-  Our internal data model itself (inclusive of our desire to not incidentally create _another_ observability data format)
+-  How we serialize that data for internal use
```

Comment:
ITYM how we serialize the data for communication with another Vector instance, since we don't use the serialized version internally otherwise.

Reply:
It is quite possible though that we'd want pick different tradeoffs for disk buffers than to the cross-instance communication. While it's an excellent high-level overview, we should probably view the encoding problem together with the expected usage patterns we expect to use it for. I.e. disk buffers can have random reads (sometimes), but flushing takes longer due to fsync (also sometimes), while the cross-instance communication, being over the network, might have different properties - like packet fragmentation (in case of non-TCP backed transport), lack of the said random reads in general, usually quicker flushes.

I guess it's absolutely nesessary to have a high-level understanding of what the different encoding formats offer, and this RFC is excellent at delivering this. My concern is that to do a leap in optimization, we might have to let go of some of the conventional abtration layers here, and look at the problem across multiple OSI layers at once.